pkg/common_controller/snapshot_controller.go

```diff
@@ -1103,124 +1125,58 @@ func (ctrl *csiSnapshotController) removeSnapshotFinalizer(snapshot *crdv1.Volum
 		klog.Errorf("failed to update snapshot store %v", err)
 	}
 
-	klog.V(5).Infof("Removed protection finalizer from volume snapshot %s", snapshotKey(snapshot))
-	return nil
-}
-
-// ensureSnapshotSourceFinalizer checks if a Finalizer needs to be added for the snapshot source;
-// if true, adds a Finalizer for VolumeSnapshot Source PVC
-func (ctrl *csiSnapshotController) ensureSnapshotSourceFinalizer(snapshot *crdv1.VolumeSnapshot) error {
-	// Get snapshot source which is a PVC
-	pvc, err := ctrl.getClaimFromVolumeSnapshot(snapshot)
-	if err != nil {
-		klog.Infof("cannot get claim from snapshot [%s]: [%v] Claim may be deleted already.", snapshot.Name, err)
-		return nil
-	}
-
-	// If PVC is not being deleted and PVCFinalizer is not added yet, the PVCFinalizer should be added.
-	if pvc.ObjectMeta.DeletionTimestamp == nil && !slice.ContainsString(pvc.ObjectMeta.Finalizers, PVCFinalizer, nil) {
-		// Add the finalizer
-		pvcClone := pvc.DeepCopy()
-		pvcClone.ObjectMeta.Finalizers = append(pvcClone.ObjectMeta.Finalizers, PVCFinalizer)
-		_, err = ctrl.client.CoreV1().PersistentVolumeClaims(pvcClone.Namespace).Update(pvcClone)
-		if err != nil {
-			klog.Errorf("cannot add finalizer on claim [%s] for snapshot [%s]: [%v]", pvc.Name, snapshot.Name, err)
-			return newControllerUpdateError(pvcClone.Name, err.Error())
-		}
-		klog.Infof("Added protection finalizer to persistent volume claim %s", pvc.Name)
-	}
-
+	klog.V(5).Infof("Removed protection finalizer from volume snapshot %s", utils.SnapshotKey(snapshot))
 	return nil
 }
 
-// removeSnapshotSourceFinalizer removes a Finalizer for VolumeSnapshot Source PVC.
-func (ctrl *csiSnapshotController) removeSnapshotSourceFinalizer(snapshot *crdv1.VolumeSnapshot) error {
-	// Get snapshot source which is a PVC
-	pvc, err := ctrl.getClaimFromVolumeSnapshot(snapshot)
-	if err != nil {
-		klog.Infof("cannot get claim from snapshot [%s]: [%v] Claim may be deleted already. No need to remove finalizer on the claim.", snapshot.Name, err)
-		return nil
-	}
-
-	pvcClone := pvc.DeepCopy()
-	pvcClone.ObjectMeta.Finalizers = slice.RemoveString(pvcClone.ObjectMeta.Finalizers, PVCFinalizer, nil)
-
-	_, err = ctrl.client.CoreV1().PersistentVolumeClaims(pvcClone.Namespace).Update(pvcClone)
-	if err != nil {
-		return newControllerUpdateError(pvcClone.Name, err.Error())
+func (ctrl *csiSnapshotCommonController) contentExists(snapshot *crdv1.VolumeSnapshot) (bool, error) {
+	if snapshot.Status.BoundVolumeSnapshotContentName == nil {
```

Comment:
is snapshot.Status == nil checking needed here?

Reply:
Added.